src/components/QuranReader/ContextMenu/components/PageBookmarkAction.tsx

```diff
@@ -0,0 +1,153 @@
+  const quranReaderStyles = useSelector(selectQuranReaderStyles, shallowEqual);
+  const bookmarkedPages = useSelector(selectBookmarkedPages, shallowEqual);
+  const mushafId = getMushafId(quranReaderStyles.quranFont, quranReaderStyles.mushafLines).mushaf;
+  const toast = useToast();
+  const { cache } = useSWRConfig();
+  const { t } = useTranslation();
+  const dispatch = useDispatch();
+  const {
+    data: bookmark,
+    isValidating: isPageBookmarkedLoading,
+    mutate,
+  } = useSWRImmutable(
+    isLoggedIn() ? makeBookmarkUrl(mushafId, Number(pageNumber), BookmarkType.Page) : null,
+    async () => {
+      const response = await getBookmark(mushafId, Number(pageNumber), BookmarkType.Page);
+      return response;
+    },
+  );
+
+  const isPageBookmarked = useMemo(() => {
+    const isUserLoggedIn = isLoggedIn();
+    if (isUserLoggedIn && bookmark) {
+      return bookmark;
+    }
+    if (!isUserLoggedIn) {
+      return !!bookmarkedPages?.[pageNumber.toString()];
+    }
+    return false;
+  }, [bookmarkedPages, bookmark, pageNumber]);
+
+  // Handle bookmark toggle action
+  const handleBookmarkAdd = () => {
+    addBookmark({
+      key: Number(pageNumber),
+      mushafId,
+      type: BookmarkType.Page,
+    })
+      .then(() => {
+        mutate();
+        toast(t('quran-reader:page-bookmarked'), {
+          status: ToastStatus.Success,
+        });
+      })
+      .catch(() => {
+        toast(t('common:error.general'), {
+          status: ToastStatus.Error,
+        });
+      });
+  };
+
+  // Handle bookmark removal
+  const handleBookmarkRemove = () => {
+    deleteBookmarkById(bookmark.id)
+      .then(() => {
+        mutate();
+        toast(t('quran-reader:page-bookmark-removed'), {
+          status: ToastStatus.Success,
+        });
+      })
+      .catch(() => {
+        // Revalidate to restore the bookmark state in case of failure
+        mutate(bookmark, { revalidate: true });
+        toast(t('common:error.general'), {
+          status: ToastStatus.Error,
+        });
+      });
+  };
+
+  const onToggleBookmarkClicked = () => {
+    logButtonClick(`context_menu_page_${isPageBookmarked ? 'un_bookmark' : 'bookmark'}`);
+
+    if (isLoggedIn()) {
+      // Optimistic update
+      if (isPageBookmarked) {
+        mutate(() => null, {
+          revalidate: false,
+        });
+      }
+
+      // Invalidate bookmarks cache
+      cache.delete(
+        makeBookmarksUrl(
+          getMushafId(quranReaderStyles.quranFont, quranReaderStyles.mushafLines).mushaf,
+        ),
+      );
+
+      if (!isPageBookmarked) {
+        handleBookmarkAdd();
+      } else {
+        handleBookmarkRemove();
+      }
+    } else {
+      // For logged-out users, use Redux to store bookmarks locally
+      dispatch(togglePageBookmark(pageNumber.toString()));
+    }
+  };
+
+  let bookmarkIcon = <Spinner />;
+  if (!isPageBookmarkedLoading) {
```

Comment:
It seems requested don't get cached in the browser using useSWR on scroll so fresh requests to check if the page is bookmarked keep being fired every time. You should open Albqarah while logged in and scroll up and down to reproduce.